src/shared/dissect-image.c

```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                if (roothash_size)
+                        *roothash_size = 0;
+                if (hashdevice)
+                        *hashdevice = NULL;
                 return 0;
         }
 
-        r = getxattr_malloc(image, "user.verity.roothash", &text, true);
-        if (r < 0) {
-                char *fn, *e, *n;
+        if (roothash) {
+                assert(roothash_size);
 
-                if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
-                        return r;
+                r = getxattr_malloc(image, "user.verity.roothash", &text, true);
+                if (r < 0) {
+                        char *fn, *e, *n;
 
-                fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
-                n = stpcpy(fn, image);
-                e = endswith(fn, ".raw");
-                if (e)
-                        n = e;
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
 
-                strcpy(n, ".roothash");
+                        fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
+                        n = stpcpy(fn, image);
+                        e = endswith(fn, ".raw");
+                        if (e)
+                                n = e;
+                        else {
+                                e = endswith(fn, ".squashfs");
+                                if (e)
+                                        n = e;
+                        }
 
-                r = read_one_line_file(fn, &text);
-                if (r == -ENOENT) {
-                        *ret = NULL;
-                        *ret_size = 0;
-                        return 0;
+                        strcpy(n, ".roothash");
+
+                        r = read_one_line_file(fn, &text);
+                        if (r == -ENOENT) {
+                                *roothash = NULL;
+                                *roothash_size = 0;
+                                return 0;
+                        }
+                        if (r < 0)
+                                return r;
                 }
+
+                r = unhexmem(text, strlen(text), &k, &l);
                 if (r < 0)
                         return r;
+                if (l < sizeof(sd_id128_t))
+                        return -EINVAL;
+
+                *roothash = TAKE_PTR(k);
+                *roothash_size = l;
         }
 
-        r = unhexmem(text, strlen(text), &k, &l);
-        if (r < 0)
-                return r;
-        if (l < sizeof(sd_id128_t))
-                return -EINVAL;
+        if (hashdevice) {
+                r = getxattr_malloc(image, "user.verity.hashdevice", hashdevice, true);
+                if (r < 0) {
+                        char *e, *n;
+
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
+
+                        *hashdevice = (char *)malloc(strlen(image) + STRLEN(".hashdevice") + 1);
```

Comment:
please do not clobber return params on failure. i.e. build the path in a separate variable, and only when you know everyting worked copy them into the the return parameters.

Also, to concat strings, use strjoin()

Reply:
I'll change to set the return only at the end. For the string, this is not really concatenating but overwriting the suffix, like it was already done for the roothash - doesn't look like strjoin would do the right thing?